src/advanced_inv.cpp

```diff
@@ -1724,60 +1748,70 @@ bool advanced_inventory::query_destination( aim_location &def )
         uistate.adv_inv_last_popup_dest = menu.ret;
         return true;
     }
     return false;
 }
 
-void advanced_inventory::remove_item( advanced_inv_listitem &sitem )
+// TODO: change `*.front()' to loop index based on `count'
+void advanced_inventory::remove_item( advanced_inv_listitem &sitem, int count )
 {
     assert( sitem.area != AIM_ALL );        // should be a specific location instead
     assert( sitem.area != AIM_INVENTORY );  // does not work for inventory
-    assert( sitem.area != AIM_WORN );       // yeah what he said
-    assert( sitem.it != nullptr );
-    bool rc = false;
-
-    auto &s = squares[sitem.area];
-    if( s.id == AIM_CONTAINER ) {
-        const auto cont = s.get_container( panes[src].in_vehicle() );
-        assert( cont != nullptr );
-        assert( !cont->contents.empty() );
-        assert( &cont->contents.front() == sitem.it );
-        cont->contents.erase( cont->contents.begin() );
-        rc = true;
-    } else if( sitem.area == AIM_WORN ) {
-        rc = g->u.takeoff( sitem.it );
-    } else if( sitem.from_vehicle ) {
-        rc = s.veh->remove_item( s.vstor, sitem.it );
-    }
-
-    if( !rc ) {
-        // TODO: Make this print an error if it fails
-        g->m.i_rem( s.pos, sitem.it );
+    assert( !sitem.items.empty() );
+    bool rc = true;
+
+    while(count > 0) {
+        auto &s = squares[sitem.area];
+        if( s.id == AIM_CONTAINER ) {
+            const auto cont = s.get_container( panes[src].in_vehicle() );
+            assert( cont != nullptr );
+            assert( !cont->contents.empty() );
+            assert( &cont->contents.front() == sitem.items.front() );
+            cont->contents.erase( cont->contents.begin() );
+        } else if( sitem.area == AIM_WORN ) {
+            rc &= g->u.takeoff( sitem.items.front() );
+        } else if( sitem.from_vehicle ) {
+            rc &= s.veh->remove_item( s.vstor, sitem.items.front() );
+        } else {
+            g->m.i_rem( s.pos, sitem.items.front() );
+        }
+        if(rc == false) {
+            break;
+        }
+        sitem.items.erase(sitem.items.begin());
+        --count;
     }
 }
 
-bool advanced_inventory::add_item( aim_location destarea, item &new_item )
+bool advanced_inventory::add_item( aim_location destarea, item &new_item, int count )
 {
     assert( destarea != AIM_ALL ); // should be a specific location instead
     bool rc = true;
-    if( destarea == AIM_INVENTORY ) {
-        g->u.i_add( new_item );
-        g->u.moves -= 100;
-        return rc;
-    } else if( destarea == AIM_WORN ) {
-        return g->u.wear_item(&new_item);
-    } else {
-        advanced_inv_area &p = squares[destarea];
-        if( panes[dest].in_vehicle() ) {
-            rc = p.veh->add_item( p.vstor, new_item );
+
+    while(count > 0) {
+        if( destarea == AIM_INVENTORY ) {
+            auto it = g->u.i_add( new_item );
+            g->u.moves -= 100;
+            rc = it.type->id == new_item.type->id;
```

Comment:
Whoa, when do you expect this to be false? `player::i_add` returns a *reference* to the added item in the inventory, it should *never* be a different item. But even if the item changed its type during pickup (why actually not?), that wouldn't mean the action failed (which is what `rc == false` would indicate).